File: reactapp/src/types/substitution/bartering/merchant.ts
import { Type } from "class-transformer";
import { ILinearEquation, LinearEquation } from "@/types/math/linearEquation.ts";
import { Trade } from "./trade.ts";

export interface IMerchant {
    readonly productType: ProductType;
    readonly linearEquation: ILinearEquation;
}

export class Merchant implements IMerchant {
    public readonly productType: ProductType;

    @Type(() => LinearEquation)
    public readonly linearEquation: LinearEquation;

    public trade!: Trade;

    constructor(productType: ProductType, linearEquation: LinearEquation) {
        this.productType = productType;
        this.linearEquation = linearEquation;
    }

    public setTrade(): void {
        this.trade = Trade.computeTradeFromEquation(this.linearEquation);
    }
}

export enum ProductType {
    Fruits,
    Goods,
    Spices
}

File: webapi/Models/Substitution/Bartering/Merchant.cs
using webapi.Models.Math;

namespace webapi.Models.Substitution.Bartering
{
    public class Merchant
    {
        public ProductType ProductType { get; set; }

        public required LinearEquation LinearEquation { get; set; }

        public Merchant() { }
    }

    public enum ProductType
    {
        Fruits,
        Goods,
        Spices
    }
}
